### PR TITLE
feat: enable pausing handling keystrokes in watch mode

### PR DESCRIPTION
**Pull request by szymonrybczak**
Summary:
---------

In this PR I created `KeyPressHandler` that allows us to pause handling keystrokes because in nearest future we will add interactive prompts to `start` command, so they will run in watch mode. If we do not pause the handling keystrokes, the "focus" will be on the prompt. I created a function that wraps `prompts` and pauses and resumes interactivity in watchMode. 

Test Plan:
----------

1. Clone the repository and do all the required steps from the [Contributing guide](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md)
2. Run this command:

```sh
node /path/to/react-native-cli/packages/cli/build/bin.js start
```
3. Press key that we're handling (`a`,` i`, `d`, `r`) and check if keystrokes are properly handled. 
 

Checklist
----------

- [x] Documentation is up to date to reflect these changes.
- [x] Follows commit message convention described in [CONTRIBUTING.md](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md#commit-message-convention)

### CODE REVIEW

File: packages/cli-plugin-metro/src/tools/KeyPressHandler.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import {CLIError, logger} from '@react-native-community/cli-tools';
+
+const CTRL_C = '\u0003';
+
+/** An abstract key stroke interceptor. */
+export class KeyPressHandler {
+  private isInterceptingKeyStrokes = false;
+  private isHandlingKeyPress = false;
+
+  constructor(public onPress: (key: string) => Promise<any>) {}
```
Comment (thymikee) [inline packages/cli-plugin-metro/src/tools/KeyPressHandler.ts:10]: I think we're don't intend to return anything from this fn and if it must be either sync or async, let's state that in the type
```suggestion
  constructor(public onPress: (key: string) => Promise<void> | void) {}
```

File: packages/cli-plugin-metro/src/commands/start/watchMode.ts [hunk 2]
```diff
@@ -37,38 +42,44 @@ function enableWatchMode(messageSocket: any, ctx: Config) {
...
-          ? ctx.project.ios?.watchModeCommandParams
-          : ctx.project.android?.watchModeCommandParams;
-      execa('npx', [
-        'react-native',
-        name === 'i' ? 'run-ios' : 'run-android',
-        ...(params ?? []),
-      ]).stdout?.pipe(process.stdout);
-    } else {
-      console.log(_key);
+  const onPressAsync = async (key: string) => {
```
[the remark this answers is not in the document]
Comment (szymonrybczak) [reply packages/cli-plugin-metro/src/commands/start/watchMode.ts:45]: we're awaiting it here: 
https://github.com/react-native-community/cli/blob/e8f7148c5d6ca71acf65b62848f05be5321f0d88/packages/cli-plugin-metro/src/tools/KeyPressHandler.ts#L40

File: packages/cli-plugin-metro/src/tools/KeyPressHandler.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    // Prevent sending another event until the previous event has finished.
+    if (this.isHandlingKeyPress && key !== CTRL_C) {
+      return;
+    }
+    this.isHandlingKeyPress = true;
+    try {
+      logger.debug(`Key pressed: ${key}`);
+      await this.onPress(key);
+    } catch (error) {
+      return new CLIError('There was an error with the key press handler.');
```
Comment (thymikee) [inline packages/cli-plugin-metro/src/tools/KeyPressHandler.ts:42]: we should pass the `error` to CLIError to get user some context. This is execa that can fail here right? Let's make sure the output is readable

File: packages/cli-plugin-metro/src/commands/start/watchMode.ts [hunk 4]
```diff
@@ -37,38 +42,44 @@ function enableWatchMode(messageSocket: any, ctx: Config) {
...
+        break;
+      case CTRL_C:
+        process.exit();
     }
-  });
+  };
+
+  const keyPressHandler = new KeyPressHandler(onPressAsync);
+  const listener = keyPressHandler.createInteractionListener();
+  addInteractionListener(listener);
```
Comment (thymikee) [inline packages/cli-plugin-metro/src/commands/start/watchMode.ts:81]: would it be a bad idea to move this call to `createInteractionListener()`?

File: packages/cli-plugin-metro/src/commands/start/watchMode.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (thymikee) [reply packages/cli-plugin-metro/src/commands/start/watchMode.ts:45]: yup, however there's no async code inside this fn so there's no point in doing that now. if the handler body becomes async, we could extend the listener type then

File: packages/cli-plugin-metro/src/tools/KeyPressHandler.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  private handleKeypress = async (key: string) => {
+    // Prevent sending another event until the previous event has finished.
+    if (this.isHandlingKeyPress && key !== CTRL_C) {
+      return;
+    }
+    this.isHandlingKeyPress = true;
+    try {
+      logger.debug(`Key pressed: ${key}`);
+      await this.onPress(key);
```
Comment (thymikee) [inline packages/cli-plugin-metro/src/tools/KeyPressHandler.ts:40]: when this gets sync (which it is right now in fact), `this.isHandlingKeyPress` gets unnecessary

**Review by thymikee — CHANGES_REQUESTED** [review]
small nits and this is good to go

File: packages/cli-plugin-metro/src/commands/start/watchMode.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (szymonrybczak) [reply packages/cli-plugin-metro/src/commands/start/watchMode.ts:45]: ah, right

**Review by thymikee — APPROVED** [review]

**Pull request merged**